src/components/Trade/hooks/useSwapper/useSwapper.ts

```diff
@@ -50,157 +54,238 @@ export enum TRADE_ERRORS {
+    }
+    return result
+  }
+
   const getQuoteFromSwapper = async <C extends ChainTypes, S extends SwapperType>({
     amount,
     sellAsset,
     buyAsset,
+    action,
     onFinish
-  }: GetQuote<C, S>) => {
+  }: GetQuoteFromSwapper<C, S>) => {
     if (debounceObj?.cancel) debounceObj.cancel()
     clearErrors()
     const quoteDebounce = debounce(async () => {
-      try {
-        const swapper = swapperManager.getSwapper(bestSwapperType)
-        let convertedAmount = amount
-        const isFiat = Object.keys(amount)[0].includes('fiat')
-        if (isFiat) {
-          const rate = await swapper.getUsdRate({
+      if (isComponentMounted.current) {
+        try {
+          const swapper = swapperManager.getSwapper(bestSwapperType)
+          let convertedAmount =
+            action === TradeActions.BUY ? { buyAmount: amount } : { sellAmount: amount }
+
+          const isFiat = action === TradeActions.FIAT
+          if (isFiat) {
+            const rate = await swapper.getUsdRate({
+              symbol: sellAsset.symbol,
+              tokenId: sellAsset.tokenId
+            })
+            convertedAmount = {
+              sellAmount: toBaseUnit(
+                bn(amount).div(rate).toString(),
+                sellAsset.precision
+              ).toString()
+            }
+          }
+          const quoteInput = {
+            sellAsset: sellAsset,
+            buyAsset: buyAsset,
+            ...convertedAmount
+          }
+
+          const { trade } = getValues()
+          let minMax = trade
+          if (
+            quote?.sellAsset?.symbol !== sellAsset.symbol &&
+            quote?.buyAsset?.symbol !== buyAsset.symbol
+          ) {
+            minMax = await swapper.getMinMax(quoteInput)
+            const minMaxTrade = { ...minMax, ...trade }
+            minMax && setValue('trade', minMaxTrade)
+          }
+          const newQuote = await swapper.getQuote({ ...quoteInput, ...minMax })
+          if (!newQuote?.success) throw newQuote
+
+          const sellAssetUsdRate = await swapper.getUsdRate({
             symbol: sellAsset.symbol,
             tokenId: sellAsset.tokenId
           })
-          const fiatAmount = Object.values(amount)[0]
-          convertedAmount = {
-            sellAmount: toBaseUnit(
-              bn(fiatAmount).div(rate).toString(),
-              sellAsset.precision
-            ).toString()
+          let buyAssetUsdRate
+          if (newQuote?.rate) {
+            buyAssetUsdRate = bn(sellAssetUsdRate).dividedBy(newQuote?.rate).toString()
+          } else {
+            buyAssetUsdRate = await swapper.getUsdRate({
+              symbol: buyAsset.symbol,
+              tokenId: buyAsset.tokenId
+            })
           }
+          const sellAssetFiatRate = bn(sellAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
+          const buyAssetFiatRate = bn(buyAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
```

Comment:
Or you can make `sellAssetUsdRate` / `buyAssetUsdRate` are BigNumbers at the top of the scope before they're used and avoid wrapping them in `bn` each time. You can then guard that it's a valid number once at the top.

```suggestion
          const sellAssetFiatRate = bnOrZerp(sellAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
          const buyAssetFiatRate = bnOrZoer(buyAssetUsdRate).times(1).toString() // TODO: Implement fiatPerUsd here
```

Reply:
Don't commit this, spelling errors.